Extension/src/SSH/sshCommandRunner.ts

```diff
@@ -0,0 +1,518 @@
+
+                            stdout = '';
+                        }
+
+                        if (interaction.canceled) {
+                            if (args.usedInteractors) {
+                                args.usedInteractors.add(interactor.id);
+                            }
+
+                            done(true);
+                            return;
+                        }
+
+                        if (interaction.continue) {
+                            if (args.usedInteractors) {
+                                args.usedInteractors.add(interactor.id);
+                            }
+                            continueWithoutExiting = true;
+                            done(false, true);
+                            return;
+                        }
+
+                        if (typeof interaction.response === 'string') {
+                            if (args.usedInteractors) {
+                                args.usedInteractors.add(interactor.id);
+                            }
+
+                            if (terminal) {
+                                terminal.sendText(`${interaction.response}\n`);
+                                const logOutput: string = interaction.isPassword
+                                    ? interaction.response.replace(/./g, '*')
+                                    : interaction.response;
+                                getOutputChannelLogger().appendLine(localize('ssh.wrote.data.to.terminal', '"{0}" wrote data to terminal: "{1}".', nickname, logOutput));
+                            }
+                        }
+                    } catch (e) {
+                        failed(e);
+                    }
+                }
+            }
+        };
+
+        if (token) {
+            tokenListener = token.onCancellationRequested(() => {
+                done(true);
+            });
+        }
+
+        const terminalIsWindows: boolean = typeof args.terminalIsWindows === 'boolean' ? args.terminalIsWindows : isWindows();
+        try {
+            // the terminal process should not fail, but exit cleanly
+            let shellArgs: string | string[];
+            if (args.sendText) {
+                shellArgs = '';
+            } else {
+                shellArgs = terminalIsWindows ? `/c ${command}\nexit /b %ErrorLevel%` : ['-c', `${command}\nexit $?`];
+            }
+
+            const options: vscode.TerminalOptions = {
+                cwd:
+                    args.cwd ||
+                    (terminalIsWindows
+                        ? vscode.Uri.file(os.homedir() || 'c:\\')
+                        : vscode.Uri.file(os.homedir() || '/')),
+                name: nickname,
+                shellPath: getShellPath(terminalIsWindows),
+                shellArgs,
+                hideFromUser: true
+            };
+
+            let terminalDataHandlingQueue: Promise<void> = Promise.resolve();
+            terminalListener = systemInteractor.onDidWriteTerminalData(async e => {
```

Comment:
You should dispose of this listener as soon as you don't need it (eg. `if (args.showLoginTerminal)`). This is because if there exists a single one of these listeners, _all_ data for _all_ terminals gets sent to _all_ extension hosts 😱

Reply:
It's disposed in clean(), at line 266. There is one case where done is called but clean isn't, which is when continueOn is specified. But in this case, the command will eventually be cancelled or the terminal will be closed, and the listeners will be disposed.